src/utils/theme.ts

```diff
@@ -0,0 +1,39 @@
+import { autoTheme, darkTheme } from "./resources";
+
+/**
+ * Emits when the theme is dynamically toggled between light and dark on <body>.
+ * This file is imported in Stencil's `globalScript` config option.
+ *
+ * @see {@link https://stenciljs.com/docs/config#globalscript Stencil's globalScript property}
+ */
+export default function (): void {
+  const { classList } = document.body;
+  const prefersDark = window.matchMedia && window.matchMedia("(prefers-color-scheme: dark)").matches;
+
+  const getTheme = (): string =>
+    classList.contains(darkTheme) || (classList.contains(autoTheme) && prefersDark) ? "dark" : "light";
+
+  const emitThemeChange = (theme: string) =>
+    document.body.dispatchEvent(new CustomEvent("calciteThemeChange", { bubbles: true, detail: { theme } }));
+
+  const themeChangeHandler = (newTheme: string): void => {
+    currentTheme !== newTheme && emitThemeChange(newTheme);
+    currentTheme = newTheme;
+  };
+
+  let currentTheme = getTheme();
+
+  // emits event on page load
+  emitThemeChange(currentTheme);
+
+  // emits event when changing OS theme preferences
+  window
+    .matchMedia("(prefers-color-scheme: dark)")
+    .addEventListener("change", (event) => themeChangeHandler(event.matches ? "dark" : "light"));
+
+  // emits event when toggling between theme classes on <body>
+  new MutationObserver(() => themeChangeHandler(getTheme())).observe(document.body, {
```

Comment:
Is it possible to add a test for this? This would be the first one we try for the global script.

Reply:
I tried creating a new file `src/utils/theme.e2e.ts` as well as adding it to an existing component's e2e tests, but neither worked. I tried w/ and w/o adding `<body>` in the HTML content. I also tried "sleeping" in case it was a timing issue:
`await new Promise((r) => setTimeout(r, 2000));`

@jcfranco / @driskull Do you notice any issues with the code below, or have any other ideas?

<details>

  <summary>Attempt 1</summary>

```ts
import { newE2EPage } from "@stencil/core/testing";
import { html } from "../../support/formatting";

it("emits calciteThemeChange when setting calcite-theme-dark class on body", async () => {
  const page = await newE2EPage();
  await page.setContent(html`<calcite-button>Test</calcite-button>`);
  const calciteThemeChangeSpy = await page.spyOnEvent("calciteThemeChange");

  document.body.classList.add("calcite-theme-dark");
  await page.waitForChanges();

  expect(document.body.classList.contains("calcite-theme-dark")).toBe(true);
  expect(calciteThemeChangeSpy).toHaveReceivedEventTimes(1);
  expect(calciteThemeChangeSpy.lastEvent.detail.theme).toBe("dark");

  document.body.classList.toggle("calcite-theme-dark");
  await page.waitForChanges();

  expect(document.body.classList.contains("calcite-theme-dark")).toBe(false);
  expect(calciteThemeChangeSpy).toHaveReceivedEventTimes(2);
  expect(calciteThemeChangeSpy.lastEvent.detail.theme).toBe("light");
});

```
![image](https://user-images.githubusercontent.com/10986395/183481539-d021f681-f973-4ab9-b668-3076cb5920db.png)

</details>

<details>

  <summary>Attempt 2</summary>

```ts
import { newE2EPage } from "@stencil/core/testing";
import { html } from "../../support/formatting";

it("emits calciteThemeChange when setting calcite-theme-dark class on body", async () => {
  const page = await newE2EPage();
  await page.setContent(html`<calcite-button>Test</calcite-button>`);

  const emitted = await page.evaluate(() => {
    const emittedEvents = [];
    document.body.addEventListener("calciteThemeChange", (event) => emittedEvents.push(event));
    document.body.classList.toggle("calcite-theme-dark");
    document.body.classList.remove("calcite-theme-dark");
    return emittedEvents;
  });

  expect(emitted).toHaveLength(2);
  expect(emitted[0].detail.theme).toBe("dark");
  expect(emitted[1].detail.theme).toBe("light");
});

```

![image](https://user-images.githubusercontent.com/10986395/183485968-0098c424-3163-4f32-9df1-aab2f298f935.png)

</details>

Reply:
I logged a ticket with stencil to see if creating an e2e test for code imported through the globalScript is possible before I spend more timing looking into this

Reply:
Update from ionic:

> The globalScript should be called as part of the newE2EPage() call, but you cannot do much direct testing of it. Your tests seem more indirect, though, so you should be OK. Looking at your PR, it seems like attempt 1 should have worked. If you can create a simplified version of this and share it, I would be happy to have a look.

I was able to [repro with the stencil starter](https://github.com/benelan/stencil-component-starter) so Ken is looking into it

Reply:
Installing this so it makes it into the release I'll be doing shortly. I'll add a test in a separate PR when Stencil gets back to us (if it's possible)